datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -827,7 +912,7 @@ mod tests {
             (
                 "2020-09-08T13:42:29.190855Z",
                 "second",
-                "2020-09-08T13:42:29.000000Z",
+                "2020-09-08T13:42:29.190855Z",
```

Comment:
This change doesn't seem correct to me -- the original value was correct (truncated to microsecond). Do you know why it was changed? As written it still has microsecond precision I think

Reply:
https://github.com/apache/arrow-datafusion/blob/b17cf0710938eed1c9f7794d5695bfac89bff580/datafusion/physical-expr/src/datetime_expressions.rs#L217-L223

Previously, the truncation of timestamps to seconds occurred prematurely in `date_trunc_single`, resulting in inaccurate results. Therefore, I have made the necessary adjustments and deferred the logic to the `_date_trunc` function. By implementing this change, the granularity of the timestamps is preserved, and truncation is performed at a later stage to ensure accuracy.

Reply:
Perhaps then we can update the test to call `_date_trunc`

```
            let result = date_trunc_single(granularity, left).unwrap();
``` 

to 
```
            let result = _date_trunc(granularity, left).unwrap();
```

As written is seems to have the wrong answer

Reply:
I modified `date_trunc_single`, so it will truncate the timestamp to second when the granularity is second.